cmd/crank/beta/validate/manager.go

```diff
@@ -201,6 +201,10 @@ func (m *Manager) addDependencies() error {
 			if len(image) > 0 {
 				image = fmt.Sprintf(imageFmt, image, dep.Version)
 				m.deps[image] = true
+
+				if _, ok := m.confs[image]; !ok && dep.Configuration != nil {
+					m.confs[image] = true
```

Comment:
it looks like this line is adding a new element to the `m.confs` map, while we are currently iterating through it (starting on line 173).  This can be problematic with undefined behavior: https://stackoverflow.com/a/68639401 

This is demonstrated in https://go.dev/play/p/V23Q7zkItEF - try running that multiple times and you see that you get inconsistent results.

Is the intent here that this iteration will also include the item we are adding here? If so, I don't think we can rely on that. What do you think of this scenario? 🤔

Reply:
With more testing now, I think this is an actual problem if there is a 3rd level of dependencies in the tree, e.g. like `xpkg.upbound.io/upbound/platform-ref-azure:v0.11.0` has.  If there are only 2 levels of dependencies, then they can be added here and it doesn't matter if the iteration misses them because they have no further dependencies to find.  However, if there is a 3rd level, and we miss a package at the 2nd level because we've added it to the map during our iteration of it, we will miss those 3rd+ level dependencies.

You can see this by running the commands in https://gist.github.com/jbw976/51d81caf0ece87e2bb6c64418ece8a8d#file-inconsistent-behavior for `platform-ref-azure` multiple times - you get an inconsistent number of dependencies found/downloaded each time.  For example:

```
❯ export CCACHE="/Users/Jared/.crossplane/cache"; rm -fr "${CCACHE}"; go run cmd/crank/main.go beta validate ~/Desktop/crossplane/test/5809/config-plat-ref-azure.yaml ~/Desktop/crossplane/test/5809/resources --cache-dir="${CCACHE}" | tee cmd-out.txt; cat cmd-out.txt | wc -l
<packages ommitted>
      15
> <run again>
      13
> <run again>
      11
```

Reply:
Hmm yes it looks like we have a bug here. I will be changing it to a new recursive approach that adds configurations and their dependencies.

Reply:
Hopefully this problematic situation is fixed with the recursive approach introduced by https://github.com/crossplane/crossplane/commit/70cc678f13cdee5e9aa36def6455f3eb762634d2